Improving the quick start documentation (#52)
peopledoc/septentrion: docs/quickstart.rst

Quickstart
==========

Welcome to the *Septentrion* quickstart documentation.

In this tutorial, you will learn how to write migrations in the expected format, how to visualize them and how to run
them against a PostgreSQL database.

Prepare your database
---------------------

If you already have a PostgreSQL database around, make sure to note the connection
parameters. Otherwise, we'll create one together with Docker_:

.. _Docker: https://docs.docker.com/

.. code-block:: console

    $ docker run --detach --rm -p 5432:5432 --name septentrion_db postgres

.. note::

    If you need to stop the docker at some point, use ``docker stop septentrion_db``.

.. note::

    If you stop the Docker or reboot your machine, the Postgres database will be trashed and you will need to re-create

>     one and add some data again.

Reviewer: If you just stopped the Postgres container you can restart it without loosing any database data. It is when you remove a container that you loose the data that was on the container file system.

Reviewer 2: I believe this is not true if it's launched as `docker run --rm` as mentioned above.

Reviewer: True. I read it too fast.

Quickstart
==========

Welcome to the *Septentrion* quickstart documentation.

In this tutorial, you will learn how to write migrations in the expected format, how to visualize them and how to run
them against a PostgreSQL database.

Prepare your database
---------------------

If you already have a PostgreSQL database around, make sure to note the connection
parameters. Otherwise, we'll create one together with Docker_:

.. _Docker: https://docs.docker.com/

.. code-block:: console

>     $ docker run --detach --rm -p 5432:5432 --name septentrion_db postgres

Reviewer 3: Perhaps, either if it's a tutorial we want a persistent database, not use the `--rm` parameter

Reviewer 2: The problem is that subsequent calls to `run` will still spawn different dockers, so running without --rm will not help a lot, afaik.

Author: We added a note to clarify the docker behaviour, so that people are not surprised when the database is empty after a reboot.

.. note::

    If you need to stop the docker at some point, use ``docker stop septentrion_db``.

> You'll also need ``psycopg2``, which is notoriously complex to install. Procrastinate

Reviewer 3: Not Procrastinate but Septentrion

Reviewer 2: Woops :)

---------------------------------

Within a virtualenv_, install Septentrion with:

.. _virtualenv: https://packaging.python.org/tutorials/installing-packages/#creating-virtual-environments

.. code-block:: console

    (venv) $ pip install septentrion

Next we will configure the connection to the PostgreSQL database. We can do this either with command line flags,
environment variables or a configuration file.
In this tutorial, we will use a configuration file, ``septentrion.ini``.

.. code-block:: ini

    [septentrion]
    host=localhost
    # port=5432
    # username=postgres

.. note::

    With the Docker setup described above, you should be good to go.
    If you need additional configuration parameters to connect to your database, have a look at
    :ref:`advanced configuration options <how-to/configure>`.

Write migrations
----------------

Migrations are SQL files that are executed in order. Migrations are grouped together in successive versions.
All migrations in the same version are placed in a folder named after the version number.

Let's create our migration folder, and a first version folder.

.. code-block:: console

    $ mkdir migrations

>     $ mkdir migrations/1.0

Reviewer 3: The default value of the setting `migration_root` is `'.'`. We have to add 'migrations_root=migrations' in the `setting.ini` file or create migrations directly on the root dir of this tutorial project.

Author: Updated `septentrion.ini`

Now we can add a migration file in ``1.0/`` named ``00-author.ddl.sql``:

.. code-block:: sql

    BEGIN;
    CREATE TABLE "author" (
        "id" serial NOT NULL PRIMARY KEY,
        "name" varchar(100) NOT NULL,
        "date_of_birth" varchar(10) NOT NULL
    );
    COMMIT;

.. note::

    Migrations are executed in alphabetical order, so an easy way to control the order is to prefix filenames with two
    digits.

.. note::

    ``ddl`` stands for *Data Definition Language* and corresponds to all the operations that change the structure of the
    database (``CREATE``, ``DROP``, ...).

Congratulations, you have now written your first *Septentrion* migration. Let's see how to run it!

Run migrations
--------------

First, we want to visualize what is going to happen, without making any change to our data yet.

.. code-block:: console

    $ septentrion --target-version 1.0 show-migrations

#TODO print expected output

Great, we can now run it for real:

.. code-block:: console

>     $ septentrion --target-version 1.0 migrate

Reviewer 3: We currently can't run the first migration wthout a schema and a version. And the schema isn't visible in the `show-migrations` output :-(

Author: Related issue for discussion https://github.com/peopledoc/septentrion/issues/54

Quickstart
==========

Use *septentrion* as your migration tool in 2 easy steps:
  1. Write your migrations.

>   2. Run your migrations.

Reviewer 2: 3. Draw the rest of the owl :D
![](https://cdn2.manhattanprep.com/gmat/wp-content/uploads/sites/18/2018/09/rm-3-image-1.png)

(note: I was the one writing this initially)

Write migrations
----------------
Your migration folder should be organised as follow:
 - each version has its own folder;
 - in a version folder, migrations will be executed in alphabetical order;

>  - `schemas` folder contains state of the schema at a given version.

Reviewer 2: ```suggestion
 - `schemas` folder contains the state of the schema at a given version.
```

> # I'm not sure if the explanation around the naming convention makes sense in a quick start, probably not?

Reviewer 2: Not sure either, especially given we're not using the convention so far. When Septentrion starts doing something with the names, it might be interesting to introduce the convention?

Let's keep it simple :)

> # Do we need something more in that section?

Reviewer 2: I guess we should be holding hands much more. Here, we're not helping people implement septentrion for their project, we're taking them on a tour showcasing septentrion. We can put ourselves in a situation where we control everything.  
```
mkdir example
cd example
virtualenv .venv
source .venv/bin/activate
mkdir 1.0
edit 1.0/<name of the first migration>.sql
<spell out exactly what to write in the migration file>
Quickly create 2 or 3 other migrations
```

File naming convention
^^^^^^^^^^^^^^^^^^^^^^

> We recommend to use the following naming convention for migration files:

Reviewer 2: ```suggestion
We recommend using the following naming convention for migration files:
```

.. code-block:: console

    [VERSION]-[MIGRATION]-[ORDER]-[TYPE].sql

 - version : the release version

 - migration : a free name for the migration

 - order : a number used to order the files inside the migration

 - type : ddl or dml, ddl if the file contains schema migration, dml if the file contains data migration.

Run migrations
--------------
Configure database connection settings
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Ensure that your database connection settings are correctly configured.

 - Either your environment variables `PGHOST`, `PGPORT` and `PGUSER` are properly set.
 - Or you can set the environment variables `SEPTENTRION_HOST`, `SEPTENTRION_PORT`, `SEPTENTRION_USERNAME`.
 - If you don't want to use environment variables, you can use the `--host`, `--port` or `--username` options

>    when running the migration.

Reviewer 2: I like when there's as little choice as possible in the tutorial part. Also, I think we should avoid assuming the user knows anything as much as possible (for instance how to setup a postgres DB), so there should be a recommanded path and a link for any other pass.

Example: https://procrastinate.readthedocs.io/en/latest/quickstart.html#prerequisites-installation

Run migrations
^^^^^^^^^^^^^^

That step boils down to type:
.. code-block:: console

    $ septentrion --target-version 1.0 migrate

in your terminal.

You should run *septentrion* in your root directory, where your migrations are.

> The final version you target is a required argument (it might change in the future).

Reviewer 2: Given we know exactly what migrations they have made, we can give them the exact command, and the exact folder.
```suggestion
The final version you target is a required argument (this might change in the future).
```

> The final version you target is a required argument (it might change in the future).

Reviewer 2: Might be interesting to show the full input and output, and explicitely say what it does (run the corresponding sql file).

A check afterwards (in PSQL, show the table structure) might be interesting to really understand what happened.

> # TODO : do we want to add something about dry-run here?

Reviewer 2: I guess launching `show migrations` before `migrate` isn't a very complicated step to add, so yeah, good idea ! :+1:

Reviewer 2: Maybe a new subsection too

> 

Reviewer 2: Can we put a conclusion sentence here, summarizing what the user did (congratulations could be nice too)
